org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/panel/RolePanel.java

```diff
@@ -686,9 +686,9 @@ public void onCallback(Integer result) {
         userPreference.setProperty(UserPreference.P_WAREHOUSE, lstItemWarehouse != null ? (String) lstItemWarehouse.getValue() : "0");
         userPreference.savePreference();
 
-        //force reload of default role
-        MRole.getDefault(m_ctx, true);
-
+        //force reload of default role when more than 1 client
+        if (lstClient.getChildren().size() > 1)
+        	MRole.getDefault(m_ctx, true);
```

Comment:
@CarlosRuiz-globalqss , the 2 changes for MRole.getDefault is not mention at all in the ticket.

Reply:
Thanks @hengsin - the query on MRole is fixed, but beyond that I analyzed what was causing that query to be so frequent if there is a cache for that.

So, found that Login.loadPreferences is calling that forcing reload (no cache) and from my tests that's not necessary at all, the cache can be used without problem and in case of changes to the role a Cache Reset will solve the problem - which is the usual case.

On the other hand, in RolePanel.validateRoles I noticed that I added that call forcing reload (no cache) because of ticket [IDEMPIERE-4268](https://idempiere.atlassian.net/browse/IDEMPIERE-4268) on the commit https://github.com/idempiere/idempiere/commit/71a2ee3
This commit says "Fix problem changing client in role panel when more than 2 clients" - so instead of doing the database query always (which is expensive) I isolated that to the case that the fix mentions, when there are more than one client (one client is the most common case for most users)
Anyways, I repeated the tests on IDEMPIERE-4268 without that cache reload and I was not able to reproduce the issue, but I left that piece of code for safety - maybe my test cases were not enough.